presto-hive/src/main/java/com/facebook/presto/hive/statistics/QuickStatsProvider.java

```diff
@@ -0,0 +1,308 @@
+                return partitionStatistics;
+            }
+            catch (InterruptedException | ExecutionException e) {
+                log.error(e, "Error while building quick stats for partition : %s", partitionId);
+                // Return empty PartitionStats for this partition
+                return empty();
+            }
+            catch (TimeoutException e) {
+                log.warn(e, "Timeout while building quick stats for partition : %s", partitionId);
+                // Return empty PartitionStats for this partition
+                return empty();
+            }
+            finally {
+                if (future.isDone()) {
+                    inProgressBuilds.remove(partitionKey);
+                }
+                else {
+                    // Remove the in-progress flag when the background fetch finishes
+                    future.whenCompleteAsync((r, e) -> inProgressBuilds.remove(partitionKey), commonPool());
+                }
+            }
+        }
+
+        // No quick stats calls were initiated, but we do have a quick stats build in progress
+        // Return empty PartitionStats for this partition, for now
+        return empty();
+    }
+
+    private PartitionStatistics buildQuickStats(String partitionKey, String partitionId, ConnectorSession session, SemiTransactionalHiveMetastore metastore, SchemaTableName table,
+            MetastoreContext metastoreContext)
+    {
+        Table resolvedTable = metastore.getTable(metastoreContext, table.getSchemaName(), table.getTableName()).get();
+        Optional<Partition> partition;
+        Path path;
+        if (UNPARTITIONED_ID.equals(partitionId)) {
+            partition = Optional.empty();
+            path = new Path(resolvedTable.getStorage().getLocation());
+        }
+        else {
+            partition = metastore.getPartition(metastoreContext, table.getSchemaName(), table.getTableName(), ImmutableList.of(partitionId));
+            path = new Path(partition.get().getStorage().getLocation());
+        }
+
+        HdfsContext hdfsContext = new HdfsContext(session, table.getSchemaName(), table.getTableName(), partitionId, false);
+        HiveDirectoryContext hiveDirectoryContext = new HiveDirectoryContext(recursiveDirWalkerEnabled ? RECURSE : IGNORED, isUseListDirectoryCache(session),
+                hdfsContext.getIdentity(), buildDirectoryContextProperties(session));
+        ExtendedFileSystem fs;
+        try {
+            fs = hdfsEnvironment.getFileSystem(hdfsContext, path);
+        }
+        catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+        // Resolve the list of files using the directory lister.
+        // If the directoryLister is a CachingDirectoryLister, this listing call will get/put the call from cache
+        ImmutableList<HiveFileInfo> fileList = ImmutableList.copyOf(directoryLister.list(fs, resolvedTable, path, partition, nameNodeStats, hiveDirectoryContext));
```

Comment:
Do we ever need all the files at once? It might be better to use the iterator API instead to collect the stats rather than aggregate the entire file list here. For large partitions you might even be able to start collecting stats before finishing listing which could improve the wall time due to pipelining.

Reply:
> 1/ Most implementations will use a caching directory lister, so the Iterator is just a list wrapper over a pre-cached file list

Given the assumption from the previous discussion that the quick stats is for new or old data, how likely is the result to be in the cache in the first place?

I don't know how common it is to have partitions with 1000s of files which is likely where and iterator optimization would be of benefit. I can see it being useful if someone has poorly optimized table though with little downside to well optimized ones

Reply:
My bad - I was mis-rembering the cache listing order. Quick stats would be the first to initiate directory listing, split loading that happens after planning will benefit from this
Updating the interace to user a vanilla Iterator :+1: